File: mobilev2/ios/Runner/NativeAdFactoryMedium.swift
import Flutter
import GoogleMobileAds
import UIKit
import google_mobile_ads

/// Native ad factory MEDIUM (~300pt height) — render khi
/// `NativeAd(factoryId: 'nativeMedium', ...)` được gọi từ Dart.
///
/// XIB: `NativeAdMediumView.xib`
class NativeAdFactoryMedium: FLTNativeAdFactory {

    func createNativeAd(
        _ nativeAd: GADNativeAd,
        customOptions: [AnyHashable: Any]? = nil
    ) -> GADNativeAdView? {
        guard let nibObjects = Bundle.main.loadNibNamed(
            "NativeAdMediumView", owner: nil, options: nil),
              let adView = nibObjects.first as? GADNativeAdView
        else {
            return nil
        }

        // Bind data
        (adView.headlineView as? UILabel)?.text = nativeAd.headline

        if let body = nativeAd.body {
            (adView.bodyView as? UILabel)?.text = body
            adView.bodyView?.isHidden = false
        } else {
            adView.bodyView?.isHidden = true
        }

        if let cta = nativeAd.callToAction {
            (adView.callToActionView as? UIButton)?.setTitle(cta, for: .normal)
            adView.callToActionView?.isHidden = false
        } else {
            adView.callToActionView?.isHidden = true
        }

        if let icon = nativeAd.icon?.image {
            (adView.iconView as? UIImageView)?.image = icon
            adView.iconView?.isHidden = false
        } else {
            adView.iconView?.isHidden = true
        }

        if let advertiser = nativeAd.advertiser {
            (adView.advertiserView as? UILabel)?.text = advertiser
            adView.advertiserView?.isHidden = false
        } else {
            adView.advertiserView?.isHidden = true
        }

        // CTA không nên auto-click bởi user gesture
        adView.callToActionView?.isUserInteractionEnabled = false

        // Bắt buộc — AdMob dùng để track click & impression
        adView.nativeAd = nativeAd

        return adView
    }
}

File: mobilev2/android/app/src/main/kotlin/com/example/flutter_base2/NativeAdFactoryMedium.kt
package com.example.flutter_base2

import android.view.LayoutInflater
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import com.google.android.gms.ads.nativead.MediaView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin

/**
 * Native ad factory MEDIUM (~300dp height) — render khi
 * `NativeAd(factoryId: 'nativeMedium', ...)` được gọi từ Dart.
 *
 * Layout: `res/layout/native_ad_medium.xml`
 */
class NativeAdFactoryMedium(
    private val layoutInflater: LayoutInflater
) : GoogleMobileAdsPlugin.NativeAdFactory {

    override fun createNativeAd(
        nativeAd: NativeAd,
        customOptions: MutableMap<String, Any>?
    ): NativeAdView {
        val view = layoutInflater.inflate(R.layout.native_ad_medium, null) as NativeAdView

        val headline = view.findViewById<TextView>(R.id.ad_headline)
        val body = view.findViewById<TextView>(R.id.ad_body)
        val cta = view.findViewById<Button>(R.id.ad_call_to_action)
        val icon = view.findViewById<ImageView>(R.id.ad_icon)
        val advertiser = view.findViewById<TextView>(R.id.ad_advertiser)
        val media = view.findViewById<MediaView>(R.id.ad_media)

        // Bind data
        headline.text = nativeAd.headline
        view.headlineView = headline

        nativeAd.body?.let {
            body.text = it
            body.visibility = android.view.View.VISIBLE
        } ?: run { body.visibility = android.view.View.GONE }
        view.bodyView = body

        nativeAd.callToAction?.let {
            cta.text = it
            cta.visibility = android.view.View.VISIBLE
        } ?: run { cta.visibility = android.view.View.GONE }
        view.callToActionView = cta

        nativeAd.icon?.let {
            icon.setImageDrawable(it.drawable)
            icon.visibility = android.view.View.VISIBLE
        } ?: run { icon.visibility = android.view.View.GONE }
        view.iconView = icon

        nativeAd.advertiser?.let {
            advertiser.text = it
            advertiser.visibility = android.view.View.VISIBLE
        } ?: run { advertiser.visibility = android.view.View.GONE }
        view.advertiserView = advertiser

        view.mediaView = media

        // Finalize — bắt buộc để AdMob track click & impression
        view.setNativeAd(nativeAd)
        return view
    }
}
